### ISSUE TITLE
Unparseable section with Databricks function and set operators

### ISSUE THREAD

**Issue by avinmathew** [sqlfluff/sqlfluff#7239]
### Search before asking

- [x] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

I receive a "Found unparsable section" when adding a set operator, such as a union or intersect, to a Databricks function.

### Expected Behaviour

Parsing should not fail when there is a union in the Databricks function, given that the code executes successfully Databricks.

### Observed Behaviour

I received the error:
```
L:   5 | P:   1 |  PRS | Line 5, Position 1: Found unparsable section: 'UNION
                       | ALL\n\nSELECT col\nFROM example_table...'
```

### How to reproduce

This is a minimal SQL will reproduce the problem with either linting or fixing:
```sql
CREATE FUNCTION example_function ()
RETURNS TABLE
RETURN
SELECT 1 AS col
UNION ALL
SELECT 2 AS col
```

Adding a FROM clause doesn't change the behaviour.

Removing the UNION works without error:
```sql
CREATE FUNCTION example_function ()
RETURNS TABLE
RETURN
SELECT 1 AS col
```

Wrapping the queries in CTEs also works without error:
```sql
CREATE FUNCTION example_function ()
RETURNS TABLE
RETURN
WITH q1 AS (
    SELECT 1 AS col
),

q2 AS (
    SELECT 2 AS col
)

SELECT * FROM q1
UNION ALL
SELECT * FROM q2;
```

### Dialect

Databricks

### Version

sqlfluff, version 3.5.0
Python 3.11.9

### Configuration

[sqlfluff]
dialect = databricks

### Are you willing to work on and submit a PR to address the issue?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [x] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)
